### PR TITLE
perf: improve the performance of using referrers tag schema

### LINKED ISSUE #354: Improve the Performance of Referrers Tag Schema
The current implementation of fallback tag scheme of the referrers API is workable but inefficient. Each time pushing a manifest with a subject requires updating the index. We can improve the performance by combining the index updates into a single operation, using the same concurrency model as demonstrated in [shizhMSFT/quay](https://github.com/shizhMSFT/quay).

In addition, we can further improve the performance of checking the capability of the referrers API by combining those requests into a single operation.

### PR DESCRIPTION
**Pull request by Wwwsylvia**
Resolves #354 
Signed-off-by: Lixia (Sylvia) Lei <lixlei@microsoft.com>

### CODE REVIEW

File: internal/syncutil/merge.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+limitations under the License.
+*/
+
+package syncutil
+
+import "sync"
+
+// mergeStatus represents the status of a merge operation.
+type mergeStatus struct {
+	// main indicates if the item is a main item.
```
Comment (shizhMSFT) [inline internal/syncutil/merge.go:22]:
```suggestion
	// main indicates if items are being merged by the current go-routine.
```

File: internal/syncutil/merge.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncutil
+
+import "sync"
+
+// mergeStatus represents the status of a merge operation.
```
Comment (shizhMSFT) [inline internal/syncutil/merge.go:20]:
```suggestion
// mergeStatus represents the merge status of an item.
```

File: internal/syncutil/merge.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+type mergeStatus struct {
+	// main indicates if the item is a main item.
+	main bool
+	// err is the error of the main item.
```
Comment (shizhMSFT) [inline internal/syncutil/merge.go:24]:
```suggestion
	// err represents the error of the merge operation.
```

File: internal/syncutil/merge.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	lock          sync.Mutex
+	committed     bool
+	items         []T
+	status        chan mergeStatus
+	pending       []T
+	pendingStatus chan mergeStatus
+}
+
+// Do calls prepare before the merge and calls resolve on the merged items.
+// If Do is called multiple times, only one of the calls will be selected to
```
Comment (shizhMSFT) [inline internal/syncutil/merge.go:64]:
```suggestion
// If Do is called multiple times concurrently, only one of the calls will be selected to
```

File: internal/syncutil/merge.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+type Merge[T any] struct {
+	lock          sync.Mutex
+	committed     bool
+	items         []T
+	status        chan mergeStatus
+	pending       []T
+	pendingStatus chan mergeStatus
+}
+
+// Do calls prepare before the merge and calls resolve on the merged items.
```
Comment (shizhMSFT) [inline internal/syncutil/merge.go:63]: Do merges concurrent operations of items into a single call of prepare and resolve.

File: internal/syncutil/merge.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		if err == nil {
+			err = resolve(items)
+		}
+		m.complete(err)
+		return err
+	}
+	return status.err
+}
+
+// assign adds a new item into the merge list.
```
Comment (shizhMSFT) [inline internal/syncutil/merge.go:80]:
```suggestion
// assign adds a new item into the item list.
```

File: internal/syncutil/merge.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+	if m.status == nil {
+		m.status = make(chan mergeStatus, 1)
+		m.status <- mergeStatus{main: true}
+	}
+	m.items = append(m.items, item)
+	return m.status
+}
+
+// commit commits the merge, and the merge is then ready for resolve.
```
Comment (shizhMSFT) [inline internal/syncutil/merge.go:101]: commit closes the assignment window after prepare, and is ready to resolve the assigned items

File: internal/syncutil/pool.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	// It may not be changed concurrently with calls to Get.
+	New func() T
+
+	lock  sync.Mutex
+	items map[any]*poolItem[T]
+}
+
+// Get gets the item identified by id.
+// The caller should invoke the returned function after using the returned item.
+func (p *Pool[T]) Get(id any) (*T, func()) {
```
Comment (shizhMSFT) [inline internal/syncutil/pool.go:39]: We may rename `id` to `key` and update related docs.

File: internal/syncutil/pool.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	// New optionally specifies a function to generate a value when Get would
+	// otherwise return nil.
+	// It may not be changed concurrently with calls to Get.
+	New func() T
+
+	lock  sync.Mutex
+	items map[any]*poolItem[T]
+}
+
+// Get gets the item identified by id.
```
Comment (shizhMSFT) [inline internal/syncutil/pool.go:37]:
```suggestion
// Get gets the value identified by key.
```

File: registry/remote/referrers.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+)
+
+// referrerOperation represents an operation on a referrer.
+type referrerOperation = int32
+
+const (
+	// referrersStateUnknown represents an addition operation on a referrer.
+	referrerOperationAdd referrerOperation = iota
+	// referrersStateUnknown represents a removal operation on a referrer.
+	referrerOperationRemove
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/referrers.go:52]: oops

File: registry/remote/repository.go [hunk 11]
```diff
@@ -1258,98 +1188,70 @@ func (s *manifestStore) indexReferrersForPush(ctx context.Context, desc ocispec.
...
-// updateReferrersIndexForPush updates the referrers index for desc referencing
-// subject on manifest push.
-func (s *manifestStore) updateReferrersIndexForPush(ctx context.Context, desc, subject ocispec.Descriptor) error {
-	// there can be multiple go-routines updating the referrers tag concurrently
+// updateReferrersIndex updates the referrers index for desc referencing subject
+// on manifest push and manifest delete.
+// References:
+//   - https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pushing-manifests-with-subject
+//   - https://github.com/opencontainers/distribution-spec/blob/main/spec.md#deleting-manifests
+func (s *manifestStore) updateReferrersIndex(ctx context.Context, desc, subject ocispec.Descriptor, operation referrerOperation) (err error) {
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/repository.go]: this is nice.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**